### ISSUE TITLE
Cell barcodes in whiteList not present in output

### ISSUE THREAD

**Issue by jrm3215** [Hoohm/CITE-seq-Count#4]
I am trying to use the CITE-seq-Count tool to associate cell hashing barcodes with cell barcodes from a 10x 3' run. I'm using the filtered barcodes.tsv (after removing '-1') as the white list to this tool. The resulting antibody/hashing barcode counts file does not contain information for every cell barcode provided. Seurat's setAssayData() requires the new assay data to have the same list of cell barcodes for both datasets. I can fill in the blanks in R after the fact but it would be nice if the CITE-seq-Count matrix had columns for every cell barcode in the white list. Am I doing something wrong?

Command:
python CITE-seq-count.py -R1 Pre_HTO_S1_L001_R1_001.fastq.gz -R2 Pre_HTO_S1_L001_R2_001.fastq.gz -t barcodes.csv -cbf 1 -cbl 16 -umif 17 -umil 26 -tr "^[ATGC]{15}[TGC][A]{6,}" -hd 1 -o Pre_HTO_antibodyCounts.txt -wl preHTO_whiteList.txt

Read 1: 26bp
Read2: 101bp

Barcodes:
GTCAACTCTTTAGCG,b1
TGATGGCCTATTGGG,b2
TTCCGCCTCTCTTTG,b3
AGTAAGTTCAGCGTA,b4

Head of white list:
TTTGTCACAATCCGAT
TTTGTCACACGGTAGA
TTTGTCACAGCTTAAC
TTTGTCAGTTAAAGTG
TTTGTCATCCGAACGC
TTTGTCATCGATAGAA
TTTGTCATCGCATGAT
TTTGTCATCTACTATC
TTTGTCATCTCTGTCG
TTTGTCATCTTGTATC

Number of cell barcodes given in white list: 8329

Number of cell barcodes present in output counts file: 7844

Thanks in advance,

Jason

**Comment by Hoohm** [comment]
Hello @jrm3215 
Could you try the following.
Take a barcode found in the 10x run but not in CITE seq and then simply grep it on R1 from the Tags data.

How does that look like? Do you find any results?

**Comment by jrm3215** [comment]
Hi @Hoohm,

Thanks for responding so quickly. You've hit on the crux of my question/issue. I've asked the tool to look for a specific set of cell barcodes and provide counts for them. It appears that when no data is found for a particular cell barcode, it is disregarded, instead of given a count of zero. If that is the intended functionality, I will fill in the zeros myself.

**Comment by Hoohm** [comment]
So what I would suggest to do, instead of filling in the zeros. First filter your mRNA data from 10x.
My guess is that some barcodes are not cells, maybe super low counts.
Because if you add 0 for the missing barcodes and then re filter it after to just dismiss the zeros you're kind of working for nothing.

But I could add a column of zeros for missing barcodes. It would allow you to filter on mRNA and CITE data at the same time, which could provide useful.

I'll add the feature.

**Comment by jrm3215** [comment]
Thanks for considering my question and potentially modifying the code. I see where you are coming from, but I think I would rather try to impute identity, rather than filter out cells when there isn't data from our low level sequencing (miSeq run) of the hashing antibody library.

**Comment by Hoohm** [comment]
@jrm3215 Wanted to get back to you and tell you it's now implemented.
Let me know if it works fine for you

**Comment by jrm3215** [comment]
The new version works fine for me. @scholzmb are you confusing hashing/cite-seq barcodes with the whitelist of cell barcodes?

**Comment by scholzmb** [comment]
You are correct.  Apologies.  This is also an issue however.  Shall I open separate issues?

**Comment by Hoohm** [comment]
Hey!
So the error you get comes from the fact that you "can't" have only 1 TAG barcode.
This is not what's really intended, the function is making sure that all the barcode TAGS are the same lenght, and don't react to having only one.

I'm working on a new implementation that will add the possibility to run different TAG barcode lengths stripping this part out.

Do you guys need to run the script with only one barcode TAG? If it's the case, I can add a quick if statement in there.
